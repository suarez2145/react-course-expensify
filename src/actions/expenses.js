import uuid from 'uuid';
import database from '../firebase/firebase';




// ADD_EXPENSE action function

// this action function is now getting its data from the stratAddExpense function below 
export const AddExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})


export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        const expense = {description, note, amount, createdAt};
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(AddExpense({
                id: ref.key,
                ...expense
            }))
        });
    }
}

// REMOVE_EXPENSE action function

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

// EDIT_EXPENSE action function

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// this action generator is the one that is responsible for retrieving the expenses data from firebase so that our 
// data is displayed and saved to the screen 

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses))
        })
    }
}