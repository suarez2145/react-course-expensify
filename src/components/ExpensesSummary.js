import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses';
import numeral from 'numeral';



export const ExpensesSummary = (props) => (

    <div>
        {
            // Here i use the ternary operator to conditionally render the paragraph with number of expenses im viewing and the total 
            props.expenses.length > 1 || props.expenses.length === 0  ? ( <p> Viewing {props.expenses.length} expenses Totaling {numeral(props.expensesTotal / 100).format('$0,0.00') } </p> )  : ( <p> Viewing {props.expenses.length} expense Totaling {numeral(props.expensesTotal / 100).format('$0,0.00') } </p> ) 
        }
        


    </div>
)

// this varaible defines what props from the state we want to pass into the new component above (ExpenseSummary)
const mapStateToProps = (state) => {
    return {
        expenses: state.expenses,
        // here i assigned the result of calling my (selectExpensesTotal) function with the expenses from my store to the property name (expensesTotal) 
        // so that i can then pass it up to my ExpensesSummary component 
        expensesTotal: selectExpensesTotal(state.expenses)

    }
}

// this connect function connects mapStateTpProps to our ExpensesSummary component
export default connect(mapStateToProps)(ExpensesSummary);