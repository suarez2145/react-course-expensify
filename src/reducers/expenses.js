



// Expenses Reducer below

// creating a variable to hold our DEFAULTs for demoState (because passing them into expenses reducers as inline arguments might get to long )
const expensesReducerDefaultState = [];


// Expenses Reducer below
const expensesReducer = (state = expensesReducerDefaultState , action) => {
    switch (action.type) {

        case 'ADD_EXPENSE':
            return [...state, action.expense]
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => {
                return id !== action.id;
            })
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id){
                    return {
                    ...expense,
                    ...action.updates
                    }

                } else {
                    return expense;
                }
            })
        default: 
        return state;
    }
}

export default expensesReducer;