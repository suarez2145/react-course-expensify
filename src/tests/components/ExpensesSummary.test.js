import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary.js';
import  expenses  from '../fixtures/expenses'; 


test('should render ExpensesSummary on DashBoardPage with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} expensesTotal={expenses}/>);
    expect(wrapper).toMatchSnapshot()
})


test('should render ExpensesSummary on DashBoardPage with a SINGLE expense', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} expensesTotal={[expenses[0]]}/>);
    expect(wrapper).toMatchSnapshot()
})