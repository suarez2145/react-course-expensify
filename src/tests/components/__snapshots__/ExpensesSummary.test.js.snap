// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary on DashBoardPage with a SINGLE expense 1`] = `
<div>
  <p>
     Viewing 
    1
     expense Totaling 
    $0.00
     
  </p>
</div>
`;

exports[`should render ExpensesSummary on DashBoardPage with multiple expenses 1`] = `
<div>
  <p>
     Viewing 
    3
     expenses Totaling 
    $0.00
     
  </p>
</div>
`;
